ykllvmwrap/src/ykllvmwrap.cc

```diff
@@ -232,6 +270,97 @@ extern "C" void *compileModule(string TraceName, Module *M,
   return ptr;
 }
 
+/// Write the string `S` in its entirety to the file descriptor `FD`.
+void writeString(int FD, string S) {
+  const char *Buf = S.c_str();
+
+  size_t Remain = strlen(Buf);
+  while (Remain) {
+    assert(Remain >= 0);
+    int Wrote = write(FD, Buf, Remain);
+    if (Wrote == -1) {
+      if (errno == EINTR)
+        continue;
+      else
+        err(EXIT_FAILURE, "write");
+    }
+
+    Remain -= Wrote;
+    Buf += Wrote;
+  }
+}
+
+/// Add debugging metadata to the module to help with debugging JITted code.
+///
+/// This works by iterating over the IR instructions of the JITted code and:
+///
+///  a) writing faked source code lines into the temporary file named by `Path`
+///     (whose file descriptor `FD` is open and ready for writing), and...
+///
+///  b) Add debug locations to the IR instructions that point to the relevant
+///     lines in the temporary file.
+///
+/// This means that debuggers (that conform to gdb's JIT interface:
+/// https://sourceware.org/gdb/current/onlinedocs/gdb/JIT-Interface.html) can
+/// show locations in the fake source code as you step over the machine code of
+/// the trace.
+///
+/// Note that the temporary file is free-form and doesn't have to be a valid
+/// source file in any particular language, so we can add any text we
+/// like to the file, if we think that would aid debugging.
+void rewriteDebugInfo(Module *M, string TraceName, int FD,
+                      const filesystem::path &Path) {
+  Function *JITFunc = M->getFunction(TraceName);
+  assert(JITFunc);
+
+  // Create a debug subprogram for the `JITFunc`.
+  DIBuilder DIB(*M);
+  DIFile *DF =
+      DIB.createFile(Path.filename().string(), Path.parent_path().string());
+  DIB.createCompileUnit(YKJIT_DWARF_LANG, DF, "ykjit", true, "", 0);
+  DISubroutineType *ST = DIB.createSubroutineType({});
+  DISubprogram *DS =
+      DIB.createFunction(DF, TraceName, TraceName, DF, 1, ST, 1,
+                         DINode::FlagZero, DISubprogram::SPFlagDefinition);
+  JITFunc->setSubprogram(DS);
+
+  // For each instruction in the trace IR, emit a human-readable version of the
+  // instruction into the temporary file and update the instruction's debug
+  // location to point to this line.
+  size_t LineNo = 1;
```

Comment:
Does `DILocation::get` start at 1?

Reply:
I will admit to being a bit suspicious of this. Does `DILocation` define itself to be 1-indexed?

Reply:
It's not documented in LLVM. It's just an `unsigned`.

If we suppose that they took precedence from DWARF, then:

> line --  An unsigned integer indicating a source line number. Lines are numbered
beginning at 1. The compiler may emit the value 0 in cases where an
instruction cannot be attributed to any source line.

(from https://dwarfstd.org/doc/DWARF4.pdf)

Reply:
I'm still a bit suspicious, but less so than before!